### PR TITLE
systemd: port to openssl 

### PR DESCRIPTION
**Pull request by kkuehlz**

### CODE REVIEW

File: src/libsystemd/sd-id128/sd-id128.c
Comment (yuwata) [inline src/libsystemd/sd-id128/sd-id128.c]: Please do not do like this. But handle error paths correctly.

File: src/basic/openssl-util.c
Comment (yuwata) [inline src/basic/openssl-util.c]: How about `return -EIO`? See `encrypt_bytes()` in homectl.c.

File: src/basic/openssl-util.c
Comment (yuwata) [inline src/basic/openssl-util.c]: Please use "macro.h"

File: src/basic/openssl-util.h
Comment (yuwata) [inline src/basic/openssl-util.h:15]: Please add a blank line between openssl/pem.h and macro.h.

File: src/basic/openssl-util.h
Comment (yuwata) [inline src/basic/openssl-util.h:17]: Please also add a blank line.

File: src/libsystemd/meson.build
Comment (keszybz) [inline src/libsystemd/meson.build]: I think it should be possible to tell meson that `libbasic_openssl` uses `openssl`, and then it will not be necessary to specify `libopenssl` separately in `dependencies`.

File: src/libsystemd/sd-id128/sd-id128.c
Comment (keszybz) [inline src/libsystemd/sd-id128/sd-id128.c]: Instead of a cast, please use `&base.bytes` and `&app_id.bytes`.

**Comment by keur** [comment]
Sorry for the no-op force push. Committed with the wrong email.

**Comment by keszybz** [comment]
LGTM, but I don't think we should merge it for this release. The problem is that this adds a new openssl dependency to libsystemd, so we'd link to both gcrypt *and* openssl. I'd very much prefer to avoid this, and line up all the changes to do the switch in one release, so that we go from depending only on gcrypt to depending only on openssl.

Another problem is with licensing. Different parties interpret things differently, but we should try to accommodate at least the more common interpretations. openssl-3 which changes the license to ASL2.0 hasn't been released yet, so we'd be linking to openssl-1 under the openssl license. Debian takes the the interpretation that the current openssl license is incompatible with gplv2, and taking FSF's dubious interpretation that mere linking causes a derivative work to be formed, means that this dependency could be problematic for projects under gplv2 linked to libsystemd. I think we should avoid this, and either wait for openssl-3 to be released, or make the dependency on openssl compile-time optional.

Would you be willing to make the implementation choice compile-time configurable? We would then be able to keep Debian happy, but still test the new code.

**Comment by keur** [comment]
@keszybz Oh boy... licenses. Anyway, I went ahead and made openssl linkage with libsystemd a compile time option!

**Comment by mbiebl** [comment]
@keszybz I very much appreciate your reply. 
I've raised this issue with a member of the Debian ftp-master team and was pointed at https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=924937

**Comment by keur** [comment]
@keszybz Please remove the merge tag. Since this is slated for the next release, I'm going to do the other openssl ports on this branch.

File: src/resolve/resolved-dns-dnssec.c
Comment (kkuehlz) [inline src/resolve/resolved-dns-dnssec.c:112]: The reason for the extra allocations is `RSA_set0_key` is the bignums get managed by openssl. There is no way to opt out of this, and kinda complicates our cleanup code. Could also get around this without the allocations with a `RSA_set0_key_TAKEPTR` wrapper function, and set m, e, d pointers to `NULL`.

We do the same thing below when duping the public key.

**Comment by keur** [comment]
@mrc0mmand If you want, you can disable CI on this branch for the time being. I am just pushing WIP stuff for the time being.

**Comment by mrc0mmand** [comment]
@keur no worries, most CIs will take care of redundant jobs, so feel free to experiment as much as you like.

**Comment by keur** [comment]
cc @poettering @keszybz @yuwata 

Can you guys take a look? I've ported a bunch of the resolve code. Could use a review before I keep going.

**Comment by keur** [comment]
@mrc0mmand If I had to guess, the bionic builds are failing because I changed importd from gcrypt to openssl, but the build machine doesn't have that dependency. From the logs it seems like only `libgnutls-openssl` is being pulled in. Anything I can do on my end?

**Comment by mrc0mmand** [comment]
> @mrc0mmand If I had to guess, the bionic builds are failing because I changed importd from gcrypt to openssl, but the build machine doesn't have that dependency. From the logs it seems like only `libgnutls-openssl` is being pulled in. Anything I can do on my end?

In that case we'll need help from @ddstreet, since he's the one responsible for Ubuntu CI and could (temporarily) add a new dependency to the autopkgbuild.

File: meson.build
Comment (mrc0mmand) [inline meson.build]:
```
WARNING: Newline character in a string detected, use ''' (three single quotes) for multiline strings instead.
This will become a hard error in a future Meson release.         message('default-dnssec cannot be set to yes or allow-downgrade when 1262 16
```

File: meson.build
Comment (mrc0mmand) [inline meson.build]: Hm, I guess with this change we'll need to update `TEST-03-JOBS` and `TEST-25-IMPORT` accordingly, since on CentOS 7 the OpenSSL is too ancient, thus `systemd-importd` is not compiled and both tests need it (at least in certain sections) to work as expected.

CentOS 7 build log:
```
Dependency openssl found: NO found 1.0.2k but need: '>= 1.1.0'
Run-time dependency openssl found: NO (tried pkgconfig and cmake)
...
Message: Skipping systemd-importd.service.8 because ENABLE_IMPORTD is false
...
disabled features: pwquality, AppArmor, openssl, libidn2, repart, homed, importd ...
```

`TEST-03-JOBS`
```
Mar 04 09:41:35 systemd-testsuite test-jobs.sh[32]: + systemctl -T start systemd-importd
Mar 04 09:41:35 systemd-testsuite systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=242 reply_cookie=0 signature=so error-na>
Mar 04 09:41:35 systemd-testsuite systemd[1]: Bus private-bus-connection: changing state RUNNING → CLOSING
Mar 04 09:41:35 systemd-testsuite systemd[1]: Bus private-bus-connection: changing state CLOSING → CLOSED
Mar 04 09:41:35 systemd-testsuite systemd[1]: Got disconnect on private connection.
Mar 04 09:41:35 systemd-testsuite systemd[1]: Bus private-bus-connection: changing state UNSET → OPENING
Mar 04 09:41:35 systemd-testsuite systemd[1]: Bus private-bus-connection: changing state OPENING → AUTHENTICATING
Mar 04 09:41:35 systemd-testsuite systemd[1]: Accepted new private connection.
Mar 04 09:41:35 systemd-testsuite systemd[1]: Bus private-bus-connection: changing state AUTHENTICATING → RUNNING
Mar 04 09:41:35 systemd-testsuite systemd[1]: Got message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=EnqueueUnitJob cookie=1 reply_cook>
Mar 04 09:41:35 systemd-testsuite systemd[1]: systemd-importd.service: Failed to load configuration: No such file or directory
Mar 04 09:41:35 systemd-testsuite systemd[1]: systemd-importd.service: Trying to enqueue job systemd-importd.service/start/replace
Mar 04 09:41:35 systemd-testsuite systemd[1]: Sent message type=error sender=org.freedesktop.systemd1 destination=n/a path=n/a interface=n/a member=n/a cookie=1 reply_cookie=1 signature=s error-name=org.freedesktop.systemd1.NoSuchUnit err>
Mar 04 09:41:35 systemd-testsuite test-jobs.sh[47]: Failed to start systemd-importd.service: Unit systemd-importd.service not found.
Mar 04 09:41:35 systemd-testsuite systemd[1]: Failed to process message type=method_call sender=n/a destination=org.freedesktop.systemd1 path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=EnqueueUnitJob cooki>
Mar 04 09:41:35 systemd-testsuite systemd[1]: systemd-importd.service: Collecting.
Mar 04 09:41:35 systemd-testsuite systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitNew cookie=2 reply_cookie=0 signat>
Mar 04 09:41:35 systemd-testsuite systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitNew cookie=243 reply_cookie=0 signature=so error-name=n>
Mar 04 09:41:35 systemd-testsuite systemd[1]: Sent message type=signal sender=org.freedesktop.systemd1 destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=3 reply_cookie=0 si>
Mar 04 09:41:35 systemd-testsuite systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.systemd1.Manager member=UnitRemoved cookie=244 reply_cookie=0 signature=so error-na>
Mar 04 09:41:35 systemd-testsuite systemd[1]: Received SIGCHLD from PID 32 (test-jobs.sh).
Mar 04 09:41:35 systemd-testsuite systemd[1]: Child 32 (test-jobs.sh) died (code=exited, status=5/NOTINSTALLED)
Mar 04 09:41:35 systemd-testsuite systemd[1]: testsuite.service: Child 32 belongs to testsuite.service.
Mar 04 09:41:35 systemd-testsuite systemd[1]: testsuite.service: Main process exited, code=exited, status=5/NOTINSTALLED
Mar 04 09:41:35 systemd-testsuite systemd[1]: testsuite.service: Failed with result 'exit-code'.
Mar 04 09:41:35 systemd-testsuite systemd[1]: testsuite.service: Service will not restart (restart setting)
Mar 04 09:41:35 systemd-testsuite systemd[1]: testsuite.service: Changed start -> failed
Mar 04 09:41:35 systemd-testsuite systemd[1]: Bus private-bus-connection: changing state RUNNING → CLOSING
Mar 04 09:41:35 systemd-testsuite systemd[1]: Sent message type=signal sender=n/a destination=n/a path=/org/freedesktop/systemd1 interface=org.freedesktop.DBus.Properties member=PropertiesChanged cookie=245 reply_cookie=0 signature=sa{sv}>
Mar 04 09:41:35 systemd-testsuite systemd[1]: Failed to send manager change signal: Connection reset by peer
Mar 04 09:41:35 systemd-testsuite systemd[1]: testsuite.service: Job 74 testsuite.service/start finished, result=failed
Mar 04 09:41:35 systemd-testsuite systemd[1]: Failed to start Testsuite service.
```

**Comment by ddstreet** [comment]
> > @mrc0mmand If I had to guess, the bionic builds are failing because I changed importd from gcrypt to openssl, but the build machine doesn't have that dependency. From the logs it seems like only `libgnutls-openssl` is being pulled in. Anything I can do on my end?
> 
> In that case we'll need help from @ddstreet, since he's the one responsible for Ubuntu CI and could (temporarily) add a new dependency to the autopkgbuild.

for ubuntu CI, openssl should be new enough in bionic, it's version 1.1.1; however the debian/control will need to be updated to list libssl-dev as a build-depends, which will need review from @mbiebl 

I opened Debian MR to track it:
https://salsa.debian.org/systemd-team/systemd/-/merge_requests/72

**Comment by keur** [comment]
I have ported all of the gcrrypt code in the tree. cc @poettering @keszybz @yuwata

File: meson.build
Comment (keszybz) [reply meson.build]: We need to retain support for gcrypt as a compile time option, because of Debian. That will solve this issue too.

File: mkosi.default
Comment (keszybz) [inline mkosi.default]: Please drop this chunk.

File: src/basic/openssl-util.c
Comment (poettering) [inline src/basic/openssl-util.c]: might as well use size_t here, given that this is memory we are talking about

File: src/basic/openssl-util.c
Comment (poettering) [inline src/basic/openssl-util.c]: `const void*` sounds more appropriate? I mean, we might want to HMAC other stuff than just plain bytes...

File: src/basic/openssl-util.c
Comment (poettering) [inline src/basic/openssl-util.c]: size_t

File: src/basic/openssl-util.c
Comment (poettering) [inline src/basic/openssl-util.c]: size_t

File: src/basic/openssl-util.c
Comment (poettering) [inline src/basic/openssl-util.c:59]: size_t

File: src/basic/openssl-util.h
Comment (poettering) [inline src/basic/openssl-util.h:28]: hmm, if we prefix this function with `openssl_`, maybe do the same for `hmac()` above?

**Comment by mbiebl** [comment]
For Debian the situation has changed in the mean time:

https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=924937#105
`consider OpenSSL, cups and libgcc as system libraries (re GPL-2) `

I very much appreciate, that @keszybz was very considerate in that regard. 
From Debian's POV there is nothing standing in the way anymore to go all OpenSSL if so desired.

Would probably be a good idea to get some input from Ubuntu on this issue as well.
@xonx, @ddstreet ^

**Comment by keszybz** [comment]
> `consider OpenSSL, cups and libgcc as system libraries (re GPL-2) `

That is great news.

> Would probably be a good idea to get some input from Ubuntu on this issue as well.

Yeah.

**Comment by poettering** [comment]
Oh, neat. THis is excellent news! If Ubuntu is fine with this, I guess we can just go OpenSSL full steam, and drop gnutls/gcrypt support bit by bit. Would love to see us remove the duplicate implementation of all this code.

**Comment by xnox** [comment]
> Oh, neat. THis is excellent news! If Ubuntu is fine with this, I guess we can just go OpenSSL full steam, and drop gnutls/gcrypt support bit by bit. Would love to see us remove the duplicate implementation of all this code.

yes it is fine. Having the standard OpenSSL exception text regardless would be nice, if at all possible.

openssl is a lot faster than gnutls/gcrypt, thus yes please use openssl / libssl / libcrypto whenever possible.

**Comment by poettering** [comment]
So I raised this on the ML too here:

https://lists.freedesktop.org/archives/systemd-devel/2020-December/045693.html

The feedback is pretty much exclusively positive.

@keur any chance you can rebase your patchset? Would you still be interested in that? That'd be awesome! Would love to get this ported over and merged. And I#d go all the way, i.e. replace gnutls/gcrypt code entirely with openssl code, so thta in the end we only have openssl as dep and no gnutls/gcrypt anymore, not even as an option.

**Review by xnox — CHANGES_REQUESTED** [review]
Unless we limit to the unreleased Apache licensed OpenSSL and all systemd code is upgradable to (l)gplv3 (via "or later" clauses); we should add OpenSSL linking exception. I.e. the one mentioned in https://people.gnome.org/~markmc/openssl-and-the-gpl.html or as specified in https://github.com/openembedded/openembedded-core/blob/master/meta/files/common-licenses/GPL-2.0-with-OpenSSL-exception

**Comment by poettering** [comment]
> Unless we limit to the unreleased Apache licensed OpenSSL and all systemd code is upgradable to (l)gplv3 (via "or later" clauses); 

All our code is lgpl2.1+, with very few exceptions. The exceptions are kernel headers we copied in which probably don#t matter. And some more liberally licensed stuff (CC0, MIT, …).

There's not "GPL2-only" or "LGPL2.1-only" code in the tree afaics.

> we should add OpenSSL linking exception. I.e. the one mentioned in https://people.gnome.org/~markmc/openssl-and-the-gpl.html or as specified in https://github.com/openembedded/openembedded-core/blob/master/meta/files/common-licenses/GPL-2.0-with-OpenSSL-exception

That means asking everyone's permission though, since that's a license change. I am too lazy for that.

Given that we only have "or later" gpl/lgpl use in our tree canonical could consider the whole project to be gpl3 licensed, no, and then the issue would go away?

File: src/basic/openssl-util.c
Comment (kkuehlz) [reply src/basic/openssl-util.c:59]: this one will conflict with openssl function signatures, but it'll be wrapped by size_t

File: src/journal/fsprg.c
Comment (kkuehlz) [inline src/journal/fsprg.c:193]: @poettering should we remove these VLAs here and elsewhere in this file? not introduced in this PR, but might be worth fixing?

Comment (keszybz) [reply src/journal/fsprg.c:193]: I think this use of VLA is appropriate. The size is limited, and the VLA is more efficient than a malloc and nicer to use than explicit alloca. But if anything, this can be discussed separately.

**Comment by poettering** [comment]
So, as I understand Debian is now friendlier to OpenSSL than Ubuntu is. If Ubuntu still isn#t OK with dropping their licensing reservations about OpenSSL I figure the situation is not as good as I though. I guess we can go fully OpenSSL only once one of the following happens:

1. Ubuntu follows Debian on this, and gives up their reservations on OpenSSL licensing
2. OpenSSL 3.0 gets finally released, which supposedly fixes the licensing problem from their side
3. we add the OpenSSL exception to our license

I don't see the third item happening though, as that means pinging every contributor, and all for something that hopefully resolves itself automatically once the second item happens.

**Comment by Maryse47** [comment]
> If Ubuntu still isn#t OK with dropping their licensing reservations about OpenSSL I figure the situation is not as good as I thought.

Did they stated this somewhere or is it pure speculation?

**Comment by poettering** [comment]
> > 
> 
> Did they stated this somewhere or is it pure speculation?

private chat with ubuntu systemd maintainers

**Comment by keszybz** [comment]
@keur sorry for the long delay.
But there is good news: openssl-3-beta1 is out, and after systemd-249 is released sometime next week or so, I want to start merging support for openssl. The bad news is that we'll need to keep it an optional requirement for now. Openssl might be out late this year, or even later, and we'll want to do a few releases ourselves before then. The choice between openssl and gcrypt (possibly with khash thrown in) will have to be made at compilation time.

How do you want to proceed with this? Any chance you could rebase and adjust this pull request?

**Comment by keur** [comment]
@keszybz I'm not sure that I guarantee the time commitment. If the idea is to maintain support for multiple crypto libraries, my suggestion would be to make a cryto util library that gets linked, and that contains all the #ifdefs. The #idfdef in `resolved` would probably become unwieldy to deal with.

**Comment by keur** [comment]
thanks @keszybz

**Pull request closed without merging**
